cpp/pgen2.h

```diff
@@ -3,73 +3,59 @@
 #ifndef CPP_PGEN2_H
 #define CPP_PGEN2_H
 
-#include "_gen/frontend/id_kind.asdl.h"
+#include <vector>
+
 #include "_gen/frontend/syntax.asdl.h"
 #include "mycpp/runtime.h"
 
-// Hacky forward declaration for translated pgen2/pnode.py
 // Note: it's probably better to express PNode in ASDL, like Token.
 namespace pnode {
-class PNode;
-}
-// Hacky stub
-namespace grammar {
-class Grammar;
-}
-
-namespace parse {
 
-class ParseError {
+class PNode {
  public:
-  ParseError(Str* msg, int type_, syntax_asdl::Token* tok)
-      : header_(obj_header()) {
-  }
-
-  static constexpr ObjHeader obj_header() {
-    return ObjHeader::ClassFixed(field_mask(), sizeof(ParseError));
-  }
-
-  static constexpr uint32_t field_mask() {
-    return maskbit(offsetof(ParseError, msg)) |
-           maskbit(offsetof(ParseError, tok));
+  PNode(int typ, syntax_asdl::Token* tok, List<PNode*>* children)
+      : typ(typ), tok(tok), children(children) {
   }
 
-  GC_OBJ(header_);
-  Str* msg;
+  int typ;
   syntax_asdl::Token* tok;
-  int type;
+  List<PNode*>* children;
 };
 
-class Parser {
- public:
-  // In C, the grammar is a constant, so the grammar arg is ignored.  (We can't
-  // get easily rid of it because the call site has to type check and run in
-  // Python.)
-  explicit Parser(grammar::Grammar* grammar) : header_(obj_header()) {
-  }
-  void setup(int start);
-  bool addtoken(int typ, syntax_asdl::Token* opaque, int ilabel);
+PNode* NewPNode(int typ, syntax_asdl::Token* tok,
+                List<pnode::PNode*>* children);
 
-  static constexpr ObjHeader obj_header() {
-    return ObjHeader::ClassFixed(field_mask(), sizeof(Parser));
-  }
+}  // namespace pnode
 
-  static constexpr uint32_t field_mask() {
-    return maskbit(offsetof(Parser, rootnode));
-  }
+namespace grammar {
 
-  // Probably should delete these
-  // void shift(int typ, syntax_asdl::Token* opaque, int newstate);
-  // void push(int typ, syntax_asdl::Token* opaque, Tuple2<List<List<Tuple2<int,
-  // int>*>*>*, Dict<int, int>*>* newdfa, int newstate);  void pop();
+typedef Tuple2<int, int> arc_t;
+typedef Dict<int, int> first_t;
+typedef List<List<arc_t*>*> states_t;
+typedef Tuple2<states_t*, first_t*> dfa_t;
+
+class Grammar {
+ public:
+  Grammar();
 
   GC_OBJ(header_);
```

Comment:
Oh BTW Chris has a change out to put the header before the object

So depending on which one lands first, we'll have to resolve, which I can help with